velox/exec/tests/utils/TestIndexStorageConnector.cpp

```diff
@@ -0,0 +1,258 @@
+  std::iota(varSizeListColumns.begin(), varSizeListColumns.end(), 0);
+  lookupResultIter_ = std::make_unique<BaseHashTable::JoinResultIterator>(
+      std::move(varSizeListColumns), 0, std::nullopt);
+  lookupResultIter_->reset(*lookupResult_);
+}
+
+std::optional<std::unique_ptr<connector::IndexSource::LookupResult>>
+TestIndexSource::ResultIterator::next(
+    vector_size_t size,
+    ContinueFuture& future) {
+  if (hasPendingRequest_.exchange(true)) {
+    VELOX_FAIL("Only one pending request is allowed at a time");
+  }
+
+  if (executor_ && !asyncResult_.has_value()) {
+    asyncLookup(size, future);
+    return std::nullopt;
+  }
+
+  SCOPE_EXIT {
+    hasPendingRequest_ = false;
+  };
+  if (asyncResult_.has_value()) {
+    VELOX_CHECK_NOT_NULL(executor_);
+    return std::move(asyncResult_.value());
+  }
+  return syncLookup(size);
+}
+
+void TestIndexSource::ResultIterator::extractLookupColumns(
+    folly::Range<char* const*> rows,
+    RowVectorPtr& result) {
+  if (result == nullptr) {
+    result = BaseVector::create<RowVector>(
+        source_->outputType_, rows.size(), source_->pool_.get());
+  } else {
+    VectorPtr output = std::move(result);
+    BaseVector::prepareForReuse(output, rows.size());
+    result = std::static_pointer_cast<RowVector>(output);
+  }
+  VELOX_CHECK_EQ(result->size(), rows.size());
+
+  for (auto projection : source_->lookupOutputProjections_) {
+    const auto resultChannel = projection.outputChannel;
+    auto& child = result->childAt(resultChannel);
+    indexTable_->table->extractColumn(rows, projection.inputChannel, child);
+  }
+}
+
+void TestIndexSource::ResultIterator::asyncLookup(
+    vector_size_t size,
+    ContinueFuture& future) {
+  VELOX_CHECK_NOT_NULL(executor_);
+  VELOX_CHECK(!asyncResult_.has_value());
+  VELOX_CHECK(hasPendingRequest_);
+
+  auto [lookupPromise, lookupFuture] =
+      facebook::velox::makeVeloxContinuePromiseContract(
+          "ResultIterator::asyncLookup");
+  future = std::move(lookupFuture);
+  auto asyncPromise = std::make_shared<facebook::velox::ContinuePromise>(
+      std::move(lookupPromise));
+  executor_->add([this, size, promise = std::move(asyncPromise)]() mutable {
+    VELOX_CHECK(!asyncResult_.has_value());
+    VELOX_CHECK(hasPendingRequest_);
+    SCOPE_EXIT {
+      hasPendingRequest_ = false;
+      promise->setValue();
+    };
+    asyncResult_ = syncLookup(size);
+  });
+}
+
+std::unique_ptr<connector::IndexSource::LookupResult>
+TestIndexSource::ResultIterator::syncLookup(vector_size_t size) {
+  VELOX_CHECK(hasPendingRequest_);
+  if (lookupResultIter_->atEnd()) {
+    return nullptr;
+  }
+
+  initBuffer<char*>(
+      source_->outputRowMapping_, source_->rawOutputRowMapping_, size);
+  initBuffer<vector_size_t>(
+      source_->inputRowMapping_, source_->rawInputRowMapping_, size);
+
+  const auto numOut = indexTable_->table->listJoinResults(
+      *lookupResultIter_,
+      /*includeMisses=*/true,
```

Comment:
Why do we need to include misses?

Reply:
This is for left join processing to identify which input is missed.